crypt.go

```diff
@@ -401,10 +402,14 @@ func activateWithTPMKey(tpm *TPMConnection, volumeName, sourceDevicePath, keyPat
 		return err
 	}
 
-	if err := activate(volumeName, sourceDevicePath, key, activateOptions); err != nil {
+	if err := activate(volumeName, sourceDevicePath, sealedKey, activateOptions); err != nil {
 		return xerrors.Errorf("cannot activate volume: %w", err)
 	}
 
+	// Ignore errors - we've activated the volume and so we shouldn't return an error at this point unless we
+	// close the volume again.
+	unix.AddKey("user", fmt.Sprintf("%s:%s:auth", filepath.Base(os.Args[0]), volumeName), authPrivateKey, userKeyring)
```

Comment:
oops, didn't quote \< properly

Reply:
I think it makes sense for having a function to retrieve data added to the kernel keyring, although I'm not sure that the API needs to be quite so complicated. I've added a "GetActivationDataFromKernel(string)" which just takes a device path and returns a piece of data with a type dependent on how the volume was activated - all of the parsing of the key description happens internally. I'm not sure what to call the function - I didn't want to use "key" in it because it returns more data than just a key - the returned type indicates how the device was unlocked and the type of key, and in the recovery case it also returns the recovery reason.

What would be the use case for having the ability to override writing the key to the kernel?

Reply:
As I said, at the moment is not even possible to log problems or control how the key is named in the kernel. Also while most snaps don't have keyring access some might (a few interfaces grant that), so it's unclear whether putting keys there is a generally viable strategy, because keyctl is not very mediation friendly atm AFAIR.